driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,62 @@ public final class CommandMessage extends RequestMessage {
         this.serverApi = serverApi;
     }
 
+    /**
+     * Create a BsonDocument representing the logical document encoded by an OP_MSG.
+     * <p>
+     * The returned document will contain all the fields from the Body (Kind 0) Section, as well as all fields represented by
+     * OP_MSG Document Sequence (Kind 1) Sections.
+     */
     BsonDocument getCommandDocument(final ByteBufferBsonOutput bsonOutput) {
-        ByteBufBsonDocument byteBufBsonDocument = ByteBufBsonDocument.createOne(bsonOutput,
-                getEncodingMetadata().getFirstDocumentPosition());
-        BsonDocument commandBsonDocument;
-
-        if (containsPayload()) {
-            commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
-
-            int payloadStartPosition = getEncodingMetadata().getFirstDocumentPosition()
-                    + byteBufBsonDocument.getSizeInBytes()
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf byteBuf = new CompositeByteBuf(byteBuffers);
+            try {
+                byteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+                ByteBufBsonDocument byteBufBsonDocument = createOne(byteBuf);
+
+                // If true, it means there is at least one Kind 1:Document Sequence in the OP_MSG
+                if (byteBuf.hasRemaining()) {
+                    BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
```

Comment:
I don't think it's a performance issue.  Consider two cases:

1.  A `findAndModify` command containing a large document.  In this case there is only a single Kind 0 section.  In this case, the check for `byteBuf.hasRemaining()` on line 129 will be false, and so `toBaseBsonDocument()` will not be called.  In this case, the optimization in  `ByteBufBsonDocument.getFirstKey` is helpful.
2. An `insert` command containing one or more large documents in a Kind 1 section.  In this case the check for `byteBuf.hasRemaining()` will be true, and  `toBaseBsonDocument()` will be called and the BsonDocument will be hydrated into a Map.  But the large documents in the Kind 1 Section will not.  The optimization in `ByteBufBsonDocument.getFirstKey` is no longer used, but that's ok.  We've avoided hydration of the large documents and only hydrated the small number of fields in the Kind 0 document.  It's true that this is relying on what we know to be true about how and when we use Kind 1 Sections, but I think that's ok.

> It seems that we can skip decoding all but the very first section with kind 0 (which in turn can be fake decoded via createOne(byteBuf), just like it's done now) when we decode for the purpose of getting the value of the first key. Moreover, that can be further simplified by just having a method that literally reads the first BSON key from ByteBufferBsonOutput starting at getEncodingMetadata().getFirstDocumentPosition().

Yes, I think you're correct about this.  I can work on this as part of this PR or we can ticket it separately.

> I was also wondering why we have to do this complex decoding with optimizations that fake decode, just to get to the command name, when we could have just propagated the command name from the place where we know it via CommandMessage to the place in InternalStreamConnection where we need it. Is it because we have MongoDatabase.runCommand, where we don't know what the command is until we encode the command?

Yes, we don't know the command name until we encode it when it's coming from runCommand, because runCommand can take any `Bson`.

The complex decoding was built not to just get the command name in `InternalStreamConnection` but rather to get the entire command in `LoggingCommandEventSender`for the purpose of publishing it in a `CommandStartedEvent`.  That's where we're really trying to avoid hydrating the whole document, when often consumers of that event just want to look at a couple of fields.

Reply:
Thank you for the explanations. I agree with everything but the very conclusion:

> We've avoided hydration of the large documents and only hydrated the small number of fields in the Kind 0 document.

yes, though compared to just decoding the first key this is still quite a lot.

> I can work on this as part of this PR or we can ticket it separately.

Let's do that separately from this PR: https://jira.mongodb.org/browse/JAVA-5553.